music_assistant/providers/audioscrobbler/__init__.py

```diff
@@ -0,0 +1,271 @@
+    return ScrobblerProvider(mass, manifest, config)
+
+
+class ScrobblerProvider(PluginProvider):
+    """Plugin provider to support scrobbling of tracks."""
+
+    _network: pylast._Network = None
+
+    def __init__(
+        self, mass: MusicAssistant, manifest: ProviderManifest, config: ProviderConfig
+    ) -> None:
+        """Initialize AudioScrobbler."""
+        super().__init__(mass, manifest, config)
+        if self.logger.level == logging.DEBUG:
+            pylast.logger.setLevel(logging.DEBUG)
+
+    def _get_network_config(self) -> dict[str, ConfigValueType]:
+        return {
+            CONF_API_KEY: self.config.get_value(CONF_API_KEY),
+            CONF_API_SECRET: self.config.get_value(CONF_API_SECRET),
+            CONF_PROVIDER: self.config.get_value(CONF_PROVIDER),
+            CONF_USERNAME: self.config.get_value(CONF_USERNAME),
+            CONF_SESSION_KEY: self.config.get_value(CONF_SESSION_KEY),
+        }
+
+    async def loaded_in_mass(self) -> None:
+        """Call after the provider has been loaded."""
+        await super().loaded_in_mass()
+
+        if self.config.get_value(CONF_SESSION_KEY):
+            self._network = _get_network(self._get_network_config())
+
+        # could be interesting for a "scrobble after X seconds" feature
+        # self.mass.subscribe(self._on_mass_queue_time_updated, EventType.QUEUE_TIME_UPDATED)
+        self.mass.subscribe(self._on_mass_queue_updated, EventType.QUEUE_UPDATED)
+        self.mass.subscribe(self._on_mass_media_item_played, EventType.MEDIA_ITEM_PLAYED)
+
+    async def _on_mass_queue_updated(self, event: MassEvent) -> None:
+        """Player has updated, update nowPlaying."""
+        if self._network is None:
+            return
+
+        queue: PlayerQueue = event.data
+        if queue.state == PlayerState.PLAYING and queue.current_item.media_type == MediaType.TRACK:
+            track: Track = queue.current_item.media_item
+            try:
+                self._network.update_now_playing(
+                    artist=track.artist_str,
+                    title=track.name,
+                    mbid=track.mbid,
+                    album=track.album.name if track.album else None,
+                )
+            except Exception as err:
+                self.logger.exception(err)
+
+    async def _on_mass_media_item_played(self, event: MassEvent) -> None:
+        """Media item has finished playing, we'll scrobble the track."""
+        if self._network is None:
+            return
+
+        item = await self.mass.music.get_item_by_uri(event.object_id)
+        if item.media_type is not MediaType.TRACK:
+            return
+
+        try:
+            track: Track = item
+            self._network.scrobble(
+                artist=track.artist_str,
+                title=track.name,
+                timestamp=time.time(),
+                mbid=track.mbid,
+                album=track.album.name if track.album else None,
+            )
```

Comment:
this is doing a blocking call - you need to either wrap the calls to this library into an executor or do the calls yourself using an async web client.

Reply:
Done, it's been a while since I worked in an eventlooped system. Combined with my lack of python experience I just saw `async` and assumed all was good 😂